### ISSUE TITLE
XSS bypass by DOM Clobbering

### ISSUE THREAD

**Issue by irsdl** [Microsoft/JSanity#7]
Here is the vector:

```
<form name="body" onmouseover="alert(1)" style="height:800px"><fieldset name="attributes"><form></form><form name="parentNode"><img id="attributes"></form></fieldset></form>
```

some more vectors can be seen here: http://www.thespanner.co.uk/2012/06/05/jslr/

The above payload was by @cgvwzq – Pepe Vila to bypass JSLR (Gareth Heyes' old toy)

**Comment by randomdross** [comment]
Hey Soroush!  Have you seen that this affects non-FF browsers?  I've confirmed it on the latest FF but haven't seen it's working elsewhere (yet).

**Comment by irsdl** [comment]
I had used Firefox and quickly used mine and Pepe's payload to be honest. Doesn't seem to work on chrome either for me but as this solution is vulnerable fundamentally against DOM clobbering it is just matter of time for someone to find a proper bypass I guess.

**Comment by cure53** [comment]
It is FF-only - like most clobbering problems.

**Comment by koto** [comment]
Well, my opinion is that the approach is flawed. You create a document fragment, innerHTML dirty node there, and attempt to sanitize _inline_ (with the TreeWalker), hoping to remove the bad nodes. Any clobbering here will potentially leave the tree in a dirty state, ready to be XSSed, when later on `.innerHTML` read is done. 

Instead, you should create a separate guaranteed-clean tree and one-by-one clone & append only sanitized nodes there.

**Comment by cure53** [comment]
Yet you have to have a clean document to begin with. Reminding of `<img name=body>` :)

**Comment by koto** [comment]
Yes. But that's straightforward, and with all approaches one needs at some point to get the unclobbered dirty tree root.

**Comment by randomdross** [comment]
> You create a document fragment, innerHTML dirty node there, and attempt to sanitize inline (with the TreeWalker), hoping to remove the bad nodes.

There's a little complexity here re the directModifySource and isolatedTargetDOM flags, but unless I'm smoking crack, jSanity is not sanitizing inline as you suggest.  There is a source doc created with document.implementation.createHTMLDocument and then elements / attributes / etc. are cherrypicked via a whitelist strategy and brought into the destination.  By default the destination DOM is the current page.

If you enable isolatedTargetDOM you effectively enable "paranoid mode," which involves maintaining a fully independent target DOM.  Iirc there were perf implications to this and I never saw a bypass this would have mitigated, so the three-DOM approach is not currently enabled by default.  We could revisit though if there's a need.

**Comment by koto** [comment]
Yup, you're right; I got confused by the logic (JS is hard ;) ). The correct approach is directModifySource = false and [this line](https://github.com/Microsoft/JSanity/blob/master/jsanity-0.2.js#L512) is crucial. Only ever insert elements that have been created from scratch.

**Comment by randomdross** [comment]
Ahh, but...  isolatedTargetDOM is false by default and thus [directModifySource is set to true](https://github.com/Microsoft/JSanity/blob/master/jsanity-0.2.js#L880).

And now I do recall having a reservation about moving isolatedTargetDOM to false by default...  We do cherrypick, but it doesn't in theory entirely eliminate the possibility that something will slip through.  Re-creating elements completely from scratch would be more conservative.

Anyway, if this worry can be substantiated with a PoC that works with isolatedTargetDOM set to false but not true, I'd be happy to enable isolatedTargetDOM by default.

**Comment by randomdross** [comment]
DOM Clobbering is addressed in Issue #5.  But if there are any remaining issues here please let me know or just file a bug.
